### PR TITLE
PSL-108 Walletnode Step3, Step4

### PR DESCRIPTION
**Pull request by bacnh8**
WIP
https://pastel.wiki/en/Architecture/OpenAPI/Sense#walletnode-1

- [ ] Step3-01/12: Implement Sense API /openapi/sense/start/{image_id}
This logic already exists in the NFT Art Registration, make it generic!!!

- [ ] Step4-01/02/03: Setup connections from WN to 3 SNs
This logic exists as the common code of goNode to encrypt Pastel network connections.

### CODE REVIEW

**Comment by a-ok123** [comment]
> WIP https://pastel.wiki/en/Architecture/OpenAPI/Sense#walletnode-1
> 
> * [ ]  Step3-01/12: Implement Sense API /openapi/sense/start/{image_id}
>   This logic already exists in the NFT Art Registration, make it generic!!!
> * [ ]  Step4-01/02/03: Setup connections from WN to 3 SNs
>   This logic exists as the common code of goNode to encrypt Pastel network connections. ~~it will require an extension for Open API cases: call back over WS to client app and asking it to sign initial handshake message~~

this is not a case anymore, see wiki

**Review by nguyen202 — DISMISSED** [review]
LGTM!

File: walletnode/api/services/sense.go [hunk 1]
```diff
@@ -115,6 +119,67 @@ func (service *Sense) ActionDetails(ctx context.Context, p *sense.ActionDetailsP
...
+// StartProcessing - Starts a processing image task
+func (service *Sense) StartProcessing(ctx context.Context, p *sense.StartProcessingPayload) (res *sense.StartProcessingResult, err error) {
+	ticket := &senseregister.Request{
+		BurnTxID:              p.BurnTxid,
+		AppPastelID:           p.AppPastelID,
+		AppPastelIDPassphrase: p.AppPastelidPassphrase,
+	}
+
+	// get image filename from storage based on image_id
+	filename, err := service.db.Get(p.ImageID)
```
Comment (mateeullahmalik) [inline walletnode/api/services/sense.go:131]: This is design layer & you are accessing db\storage layer from design layer which is not the way to go. 
Please review code of artwork register (or other APIs) & follow the same approach.

Comment (mateeullahmalik) [reply walletnode/api/services/sense.go:131]: db\storage layer and other services should be accessed from application layer.

File: walletnode/api/services/sense.go [hunk 2]
```diff
@@ -115,6 +119,67 @@ func (service *Sense) ActionDetails(ctx context.Context, p *sense.ActionDetailsP
...
+
+	// get image data from storage
+	file, err := service.register.Storage.File(string(filename))
+	if err != nil {
+		return nil, sense.MakeInternalServerError(err)
+	}
+	ticket.Image = file
+
+	// Create task
+	taskID := service.register.AddTask(ticket)
```
Comment (mateeullahmalik) [inline walletnode/api/services/sense.go:144]: first, add task.
and the task will do all the stuff for you and provide status updates. 
again, see artwork register.

File: walletnode/services/senseregister/task.go [hunk 3]
```diff
@@ -40,8 +46,123 @@ func (task *Task) Run(ctx context.Context) error {
...
+	err = task.validateMNsInfo(ctx, topNodes)
+	if err != nil {
+		return errors.Errorf("validate MNs info: %v", err)
+	}
+
+	return nil
+}
+
+// pastelTopNodes - get 10 top MNs from pastel network
+func (task *Task) pastelTopNodes(ctx context.Context) (node.List, error) {
```
Comment (mateeullahmalik) [inline walletnode/services/senseregister/task.go:88]: this code already exists... in artwork register I believe... can we use on all sides from a common place?

**Review by mateeullahmalik — CHANGES_REQUESTED** [review]
Please fix the issues mentioned.

File: walletnode/api/services/sense.go (hunk 1 above)
Comment (bacnh8) [reply walletnode/api/services/sense.go:131]: db  storage.KeyValue - in this layer to store image filename + ID. Same as artwork layer:
https://github.com/pastelnetwork/gonode/blob/9394a8bc59af9826d1ce6e8ea5369ac3028ce6f1/walletnode/api/services/artwork.go#L106

**Pull request merged**